# Системы хранения и обработки данных. HW №3

**Тема домашней работы** - группировка данных и оконные функции.

**Цель домашней работы** - научиться группировать и агрегировать данные, считать на их основе множество показателей.

In [2]:
# импортируем библиотеки для выполнения проекта
import pandas as pd
import psycopg2 as pg

In [21]:
pd.set_option('display.max_columns', None)

# Создание таблиц

1. Создать таблицы со следующими структурами и загрузить данные из csv-файлов. Детали приведены ниже. Это задание делали в предыдущем домашнем задании и будем использовать существующую БД из ДЗ №2

![Создание таблицы Customer](images/create_table_customer.png)

![Создание таблицы Trabsaction](images/create_table_transaction.png)

## Создание запросов

In [4]:
# Подключение к базе данных
conn = pg.connect(dbname="project_3_db", host="localhost", user="postgres", password="23070108", port=5432)
# Создание курсора
cursor = conn.cursor()

Вывести распределение (количество) клиентов по сферам деятельности, отсортировав результат по убыванию количества. — (1 балл)

In [5]:
# SQL-запрос
sql_query = """
SELECT job_industry_category, COUNT(*) AS count
FROM customer
GROUP BY job_industry_category
ORDER BY count DESC;
"""

# Загрузка данных в DataFrame
df = pd.read_sql_query(sql_query, conn)

# Вывод DataFrame
display(df)

/var/folders/01/wgcp1v9n5zjdx14pvmwxt4mc0000gn/T/ipykernel_9214/3810373753.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(sql_query, conn)


job_industry_category  count
0         Manufacturing    799
1    Financial Services    774
2                   n/a    656
3                Health    602
4                Retail    358
5              Property    267
6                    IT    223
7         Entertainment    136
8           Argiculture    113
9    Telecommunications     72

Найти сумму транзакций за каждый месяц по сферам деятельности, отсортировав по месяцам и по сфере деятельности. — (1 балл)

In [9]:
# SQL-запрос
sql_query = """
SELECT EXTRACT(MONTH FROM t.transaction_date) AS month,
       c.job_industry_category,
       SUM(t.list_price) AS total_transaction_amount
FROM transaction t
JOIN customer c ON t.customer_id = c.customer_id
GROUP BY EXTRACT(MONTH FROM t.transaction_date), c.job_industry_category
ORDER BY EXTRACT(MONTH FROM t.transaction_date), c.job_industry_category;
"""

# Загрузка данных в DataFrame
df = pd.read_sql_query(sql_query, conn)

# Вывод DataFrame
display(df)

/var/folders/01/wgcp1v9n5zjdx14pvmwxt4mc0000gn/T/ipykernel_9214/2900651560.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(sql_query, conn)


month job_industry_category  total_transaction_amount
0      1.0           Argiculture                  12908.83
1      1.0         Entertainment                  11139.99
2      1.0    Financial Services                  65941.48
3      1.0                Health                  51659.41
4      1.0                    IT                  30992.91
5      1.0         Manufacturing                  53925.01
6      1.0              Property                  21459.48
7      1.0                Retail                  37090.90
8      1.0    Telecommunications                   7562.31
9      1.0                   n/a                  54775.92
10     2.0           Argiculture                  11251.61
11     2.0         Entertainment                  13018.21
12     2.0    Financial Services                  68218.43
13     2.0                Health                  57108.97
14     2.0                    IT                  15991.33
15     2.0         Manufacturing                  65897.69
16     2.0              Property                  18027.83
17     2.0                Retail                  28029.71
18     2.0    Telecommunications                   7188.15
19     2.0                   n/a                  48208.01
20     3.0           Argiculture                   7018.20
21     3.0         Entertainment                  15505.97
22     3.0    Financial Services                  51471.70
23     3.0                Health                  46801.81
24     3.0                    IT                  10917.66
25     3.0         Manufacturing                  65552.97
26     3.0              Property                  24701.58
27     3.0                Retail                  44205.72
28     3.0    Telecommunications                   7120.21
29     3.0                   n/a                  58238.86
30     4.0           Argiculture                   5224.01
31     4.0         Entertainment                  15798.78
32     4.0    Financial Services                  57390.24
33     4.0                Health                  43068.42
34     4.0                    IT                  21588.27
35     4.0         Manufacturing                  75997.62
36     4.0              Property                  18939.89
37     4.0                Retail                  33127.42
38     4.0    Telecommunications                   8368.51
39     4.0                   n/a                  53501.05
40     5.0           Argiculture                   7303.41
41     5.0         Entertainment                  16364.97
42     5.0    Financial Services                  73531.99
43     5.0                Health                  54853.17
44     5.0                    IT                  14450.97
45     5.0         Manufacturing                  78669.74
46     5.0              Property                  23260.79
47     5.0                Retail                  24780.64
48     5.0    Telecommunications                   8490.16
49     5.0                   n/a                  57801.80
50     6.0           Argiculture                   4042.56
51     6.0         Entertainment                  11810.86
52     6.0    Financial Services                  71485.67
53     6.0                Health                  42369.86
54     6.0                    IT                  21697.84
55     6.0         Manufacturing                  73092.80
56     6.0              Property                  13553.36
57     6.0                Retail                  23985.48
58     6.0    Telecommunications                   3964.41
59     6.0                   n/a                  52311.64
60     7.0           Argiculture                   7065.49
61     7.0         Entertainment                   7971.33
62     7.0    Financial Services                  58895.97
63     7.0                Health                  69857.26
64     7.0                    IT                  16039.12
65     7.0         Manufacturing                  67110.37
66     7.0              Property                  23

Вывести количество онлайн-заказов для всех брендов в рамках подтвержденных заказов клиентов из сферы IT. — (1 балл)

In [15]:
# SQL-запрос
sql_query = """
SELECT t.brand, COUNT(*) AS online_orders_count
FROM transaction t
JOIN customer c ON t.customer_id = c.customer_id
WHERE c.job_industry_category = 'IT'
AND t.order_status = 'Approved'
AND t.online_order = True
GROUP BY t.brand
ORDER BY online_orders_count DESC;
"""

# Загрузка данных в DataFrame
df = pd.read_sql_query(sql_query, conn)

# Вывод DataFrame
display(df)

/var/folders/01/wgcp1v9n5zjdx14pvmwxt4mc0000gn/T/ipykernel_9214/70297620.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(sql_query, conn)


brand  online_orders_count
0           Solex                   20
1      OHM Cycles                   19
2  Norco Bicycles                   18
3  Giant Bicycles                   16
4   Trek Bicycles                   14
5        WeareA2B                   12
6                                    3

Найти по всем клиентам сумму всех транзакций (list_price), максимум, минимум и количество транзакций, отсортировав результат по убыванию суммы транзакций и количества клиентов. Выполните двумя способами: используя только group by и используя только оконные функции. Сравните результат. — (2 балла)

In [22]:
# SQL-запрос с GROUP BY
sql_query = """
SELECT 
    c.customer_id,
    SUM(t.list_price) AS total_transactions_amount,
    MAX(t.list_price) AS max_transaction_amount,
    MIN(t.list_price) AS min_transaction_amount,
    COUNT(t.transaction_id) AS transaction_count
FROM 
    customer c
JOIN 
    transaction t ON c.customer_id = t.customer_id
GROUP BY 
    c.customer_id
ORDER BY 
    total_transactions_amount DESC,
    transaction_count DESC;
"""

# Загрузка данных в DataFrame
df = pd.read_sql_query(sql_query, conn)

# Вывод DataFrame
display(df)

/var/folders/01/wgcp1v9n5zjdx14pvmwxt4mc0000gn/T/ipykernel_9214/3921676332.py:21: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(sql_query, conn)


customer_id  total_transactions_amount  max_transaction_amount  \
0            1597                    7982.65                 1807.45   
1            2565                    7860.82                 2083.94   
2            2309                    7791.91                 2005.66   
3             338                    7027.28                 1720.70   
4             799                    7004.32                 1635.30   
...           ...                        ...                     ...   
2307         1053                      12.01                   12.01   
2308         3110                      12.01                   12.01   
2309          295                      12.01                   12.01   
2310         2811                      12.01                   12.01   
2311          150                      12.01                   12.01   

      min_transaction_amount  transaction_count  
0                    1466.68                  5  
1                     183.86                  7  
2                     290.62                  6  
3                    1172.78                  5  
4                    1163.89                  5  
...                      ...                ...  
2307                   12.01                  1  
2308                   12.01                  1  
2309                   12.01                  1  
2310                   12.01                  1  
2311                   12.01                  1  

[2312 rows x 5 columns]

В этом запросе:
- Мы объединяем таблицы customer и transaction по customer_id.
- Затем мы используем оператор GROUP BY для группировки результатов по customer_id.
- Для каждого клиента вычисляется сумма всех транзакций (total_transactions_amount), максимум (max_transaction_amount) и минимум (min_transaction_amount) стоимости транзакций, а также количество транзакций (transaction_count).
- Результат сортируется сначала по убыванию суммы транзакций (total_transactions_amount) и затем по убыванию количества транзакций (transaction_count).

In [23]:
# SQL-запрос с применением оконных функций
sql_query = """
SELECT 
    customer_id,
    SUM(list_price) OVER (PARTITION BY customer_id) AS total_transactions_amount,
    MAX(list_price) OVER (PARTITION BY customer_id) AS max_transaction_amount,
    MIN(list_price) OVER (PARTITION BY customer_id) AS min_transaction_amount,
    COUNT(transaction_id) OVER (PARTITION BY customer_id) AS transaction_count
FROM 
    transaction
ORDER BY 
    total_transactions_amount DESC,
    transaction_count DESC;
"""

# Загрузка данных в DataFrame
df = pd.read_sql_query(sql_query, conn)

# Вывод DataFrame
display(df)

/var/folders/01/wgcp1v9n5zjdx14pvmwxt4mc0000gn/T/ipykernel_9214/117322177.py:17: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(sql_query, conn)


customer_id  total_transactions_amount  max_transaction_amount  \
0            1597                    7982.65                 1807.45   
1            1597                    7982.65                 1807.45   
2            1597                    7982.65                 1807.45   
3            1597                    7982.65                 1807.45   
4            1597                    7982.65                 1807.45   
...           ...                        ...                     ...   
3816          150                      12.01                   12.01   
3817          295                      12.01                   12.01   
3818         2811                      12.01                   12.01   
3819          495                      12.01                   12.01   
3820          382                      12.01                   12.01   

      min_transaction_amount  transaction_count  
0                    1466.68                  5  
1                    1466.68                  5  
2                    1466.68                  5  
3                    1466.68                  5  
4                    1466.68                  5  
...                      ...                ...  
3816                   12.01                  1  
3817                   12.01                  1  
3818                   12.01                  1  
3819                   12.01                  1  
3820                   12.01                  1  

[3821 rows x 5 columns]

В этом запросе:
- Мы используем оконные функции SUM(), MAX(), MIN() и COUNT() для вычисления суммы, максимума, минимума и количества транзакций для каждого клиента.
- Оператор PARTITION BY используется для разделения результатов на группы по customer_id.
- Результаты не группируются с помощью GROUP BY, но оконные функции применяются ко всем записям в соответствующих окнах (группах), созданных PARTITION BY.
- Результат сортируется так же, как и в первом запросе, по убыванию суммы транзакций (total_transactions_amount) и затем по убыванию количества транзакций (transaction_count).

Оба запроса возвращают аналогичные результаты, но использование оконных функций может быть более гибким, особенно когда требуется выполнить вычисления на уровне группы, но также сохранить информацию о каждой отдельной строке. В нашем случае в первом запросе из-за группировке мы получили 2312 строк, а во втором 3821 строка.

Найти имена и фамилии клиентов с минимальной/максимальной суммой транзакций за весь период (сумма транзакций не может быть null). Напишите отдельные запросы для минимальной и максимальной суммы. — (2 балла)

In [26]:
# SQL-запрос для минимальной транзакции
sql_query = """
SELECT
    c.first_name,
    c.last_name,
    SUM(t.list_price) AS total_transactions_amount
FROM
    customer c
JOIN
    transaction t ON c.customer_id = t.customer_id
GROUP BY
    c.first_name,
    c.last_name
HAVING
    SUM(t.list_price) = (
        SELECT
            MIN(sum_list_price)
        FROM
            (SELECT
                SUM(list_price) AS sum_list_price
            FROM
                transaction
            GROUP BY
                customer_id) AS subquery
    );
"""

# Загрузка данных в DataFrame
df = pd.read_sql_query(sql_query, conn)

# Вывод DataFrame
display(df)

/var/folders/01/wgcp1v9n5zjdx14pvmwxt4mc0000gn/T/ipykernel_9214/3946241721.py:29: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(sql_query, conn)


first_name  last_name  total_transactions_amount
0       Davon  Stanyland                      12.01
1      Lucius   Prescott                      12.01
2   Gabrielle   Giraudot                      12.01
3      Issiah    Fardell                      12.01
4        Lena      Chape                      12.01
5       Nevsa      Taill                      12.01
6      Bradly     Crosse                      12.01
7       Anica   Halfhyde                      12.01
8      Fabien  Iacapucci                      12.01
9      Howard    Oseland                      12.01
10     Sharia    Allkins                      12.01

In [27]:
# SQL-запрос для максимальной транзакции
sql_query = """
SELECT
    c.first_name,
    c.last_name,
    SUM(t.list_price) AS total_transactions_amount
FROM
    customer c
JOIN
    transaction t ON c.customer_id = t.customer_id
GROUP BY
    c.first_name,
    c.last_name
HAVING
    SUM(t.list_price) = (
        SELECT
            MAX(sum_list_price)
        FROM
            (SELECT
                SUM(list_price) AS sum_list_price
            FROM
                transaction
            GROUP BY
                customer_id) AS subquery
    );
"""

# Загрузка данных в DataFrame
df = pd.read_sql_query(sql_query, conn)

# Вывод DataFrame
display(df)

/var/folders/01/wgcp1v9n5zjdx14pvmwxt4mc0000gn/T/ipykernel_9214/2135123479.py:29: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(sql_query, conn)


first_name last_name  total_transactions_amount
0     Jeffry    Slowly                    7982.65

Вывести только самые первые транзакции клиентов. Решить с помощью оконных функций. — (1 балл)

In [28]:
# SQL-запрос
sql_query = """
SELECT *
FROM (
    SELECT 
        *,
        ROW_NUMBER() OVER (PARTITION BY customer_id ORDER BY transaction_date) AS transaction_rank
    FROM 
        transaction
) AS ranked_transactions
WHERE 
    transaction_rank = 1;
"""

# Загрузка данных в DataFrame
df = pd.read_sql_query(sql_query, conn)

# Вывод DataFrame
display(df)

/var/folders/01/wgcp1v9n5zjdx14pvmwxt4mc0000gn/T/ipykernel_9214/103511066.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(sql_query, conn)


transaction_id  product_id  customer_id transaction_date  online_order  \
0               9785          72            1       2017-01-05         False   
1              19498           5            3       2017-09-19         False   
2              19993          13            5       2017-04-28         False   
3              18369          72            7       2017-02-18          True   
4              19844           3            8       2017-01-28         False   
...              ...         ...          ...              ...           ...   
2307           18035           0         3495       2017-01-13          True   
2308            9769           5         3496       2017-03-07         False   
2309           17866          98         3498       2017-02-10          True   
2310           18403          67         3499       2017-11-09          True   
2311           19836          40         3500       2017-03-14          True   

     order_status           brand product_line product_class product_size  \
0        Approved  Norco Bicycles     Standard        medium       medium   
1        Approved   Trek Bicycles     Mountain           low       medium   
2        Approved           Solex     Standard        medium       medium   
3        Approved  Norco Bicycles     Standard        medium       medium   
4        Approved   Trek Bicycles     Standard        medium        large   
...           ...             ...          ...           ...          ...   
2307     Approved  Norco Bicycles         Road        medium       medium   
2308     Approved   Trek Bicycles     Mountain           low       medium   
2309     Approved   Trek Bicycles     Standard          high       medium   
2310     Approved  Norco Bicycles         Road        medium       medium   
2311     Approved      OHM Cycles     Standard          high       medium   

      list_price  standard_cost  transaction_rank  
0         360.40         270.30                 1  
1         574.64         459.71                 1  
2        1163.89         589.27                 1  
3         360.40         270.30                 1  
4        2091.47         388.92                 1  
...          ...            ...               ...  
2307      544.05         376.84                 1  
2308      574.64         459.71                 1  
2309      358.39         215.03                 1  
2310      544.05         376.84                 1  
2311     1458.17         874.90                 1  

[2312 rows x 13 columns]

Вывести имена, фамилии и профессии клиентов, между транзакциями которых был максимальный интервал (интервал вычисляется в днях) — (2 балла).

In [30]:
# SQL-запрос
sql_query = """
WITH transaction_gaps AS (
    SELECT
        customer_id,
        transaction_date,
        LAG(transaction_date) OVER (PARTITION BY customer_id ORDER BY transaction_date) AS prev_transaction_date,
        transaction_date - LAG(transaction_date) OVER (PARTITION BY customer_id ORDER BY transaction_date) AS gap
    FROM
        transaction
),
max_gap AS (
    SELECT
        MAX(gap) AS max_gap
    FROM
        transaction_gaps
)
SELECT
    c.first_name,
    c.last_name,
    c.job_title,
    mg.max_gap
FROM
    customer c
JOIN
    transaction_gaps tg ON c.customer_id = tg.customer_id
CROSS JOIN
    max_gap mg
WHERE
    tg.gap = mg.max_gap
ORDER BY
    c.first_name,
    c.last_name;
"""

# Загрузка данных в DataFrame
df = pd.read_sql_query(sql_query, conn)

# Вывод DataFrame
display(df)

/var/folders/01/wgcp1v9n5zjdx14pvmwxt4mc0000gn/T/ipykernel_9214/120642011.py:37: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(sql_query, conn)


first_name last_name        job_title  max_gap
0     Catlin     Storm  Health Coach II      344